### ISSUE TITLE
Error exit code on success

### ISSUE THREAD

**Issue by susuhahnml** [ML-KULeuven/problog#82]
Hi! I am getting an exit status code of 1 for sucess. Is this intended? I am running your example:
```
% Probabilistic facts:
0.5::heads1.
0.6::heads2.

% Rules:
twoHeads :- heads1, heads2.

% Queries:
query(heads1).
query(heads2).
query(twoHeads).
```

And just running:
```
>>> problog coin.pl -v
[INFO] Output level: INFO
[INFO] Propagating evidence: 0.0000s
[INFO] Grounding: 0.0016s
[INFO] Cycle breaking: 0.0001s
[INFO] Clark's completion: 0.0000s
[INFO] DSharp compilation: 0.6023s
[INFO] Total time: 0.6084s
  heads1:	0.5
  heads2:	0.6
twoHeads:	0.3

>>> echo $?
1
```

**Comment by Mezzenilium** [comment]
You are right, Problog always seems to end with the code: 1, whether the source of the interpreted program is correct or flawed.
I noticed this on my local version of Problog installed on a Linux PC.

**Program with error :**
(False, ParseError('% Probabilistic facts:\n0.5::heads1.\n0.6::heads2.\n\n% Rules:\ntwoHeads :- heads1, heads2\n\n% Queries:\nquery(heads1).\nquery(heads2).\nquery(twoHeads).\n', 'Expected binary operator', 98))
------------------
(program exited with code: 1)
Press return to continue
----------------------------------------------------------------------------------------------------------------------------------------

**Program without error :**
(True, {heads2: 0.6000000000000001, heads1: 0.5000000000000001, twoHeads: 0.30000000000000004})
------------------
(program exited with code: 1)
Press return to continue

The convention that a function should return 0 when its execution has been nominal has its origin in the C language. Indeed Problog seems to not respect this convention of behaviour.

In fact even when I intentionally introduced a syntax error into the code of the program written in Problog, the behaviour of the Problog interpreter was correct because it detected and reported the presence of this error.

Personally, I have not yet encountered a case where Problog does not do what it is supposed to do.

**Comment by VincentDerk** [comment]
This should be fixed in version v2.2.3 for the probability task (relevant commit b719042d5eeaf200f133440018c145b31428e3f8). 

For the other tasks (mpe, sample, ...), it is still on the internal to do list.

**Comment by sdemjanenko** [comment]
Instead of having the CLI script call `sys.exist(main())`, propagating anything `main()` returns as the exit code, it may be easier to maintain if `sys.exit` is called with a non-zero code in an `except` block. If one wants to control the non-zero code, then perhaps one can raise a specific class of exception with the desired code set.

**Comment by rmanhaeve** [comment]
Status update, since this is still open. Testing the exit code of each task on current `master` (via the console-script semantics, `sys.exit(main())`):

| invocation | exit code | |
|---|---|---|
| `problog model.pl`, success | 0 | correct |
| `problog model.pl`, parse error / missing file | 1 | correct |
| `problog mpe model.pl`, success | **1** | wrong |
| `problog sample model.pl`, success | **1** | wrong |
| unknown task name | **0** | wrong |

So the fix in 2.2.3 for the probability task still holds, but `mpe` and `sample` continue to report failure on success, and an unrecognised task name silently reports success.

Separately, the `problog-cli.py` script in the repository calls `main(...)` without `sys.exit(...)`, so it discards the return value entirely and always exits 0 — only the installed `problog` entry point propagates anything at all.

@sdemjanenko's suggestion is the right shape: rather than threading a return value through every task, exit non-zero from an exception handler and let successful paths fall through to 0. Keeping this open and labelling it a bug.

**Comment by rmanhaeve** [comment]
Fixed on the `2.2.11-fixes` branch (21d8f2d), and it turned out to be broader than this issue described: every task except the probability one reported failure on success, not just `mpe` and `sample`.

```
task     success  error
prob        0       1     (fixed in 2.2.3)
mpe         1       1
sample      1       1
ground      1       1
explain     1       1
dt          1       1
map         1       1
bn          1       1
```

They all returned the `(success, result)` pair their result handlers take, which `sys.exit()` renders as a failure. Rather than change the return value of each task — callers and the test suite rely on it — the translation now happens once in `run_task()`. @sdemjanenko, this also does what you suggested: a task that lets an exception escape now produces the same error message the tasks print themselves, instead of a traceback.

Four related problems turned up while fixing it:

- `problog-cli.py` called `main(...)` without `sys.exit(...)`, so it discarded the code entirely and always exited 0. Only the installed `problog` entry point propagated anything.
- The probability task overwrote its exit code once per input file, so `problog broken.pl fine.pl` exited 0 — a failure was masked by any later file that succeeded.
- `explain` only recorded `SUCCESS` in its result when `--web` was given.
- `ground` and `bayesnet` called `sys.exit()` from inside library code, which made them unusable programmatically as soon as anything went wrong.

All eight tasks now exit 0 on success and non-zero on failure, with regression tests covering the translation, the per-task codes, and the multi-file case.
